### PR TITLE
S 1699 tap salesforce error 431

### PR DESCRIPTION
**Pull request by JingLin0**

### CODE REVIEW

File: tap_salesforce/__init__.py [hunk 1]
```diff
@@ -51,19 +51,21 @@ def main_impl():
     stream = Stream(args.state)
     sync_fields(sf, stream)
 
-    for stream_id, fields, replication_key in sf.get_tables():
+    for table, fields, replication_key in sf.get_tables():
         if not fields:
             LOGGER.info(
-                f"skipping stream {stream_id} since it does not exist on this account"
+                f"skipping stream {table.name} since it does not exist on this account"
```
Comment (wizhi) [inline tap_salesforce/__init__.py:57]: This message seems a little misleading: wouldn't this indicate, that there are simply no fields for the table?

I see that table existence is handled in the `get_tables` function.

File: tap_salesforce/__init__.py [hunk 2]
```diff
@@ -51,19 +51,21 @@ def main_impl():
...
             )
             continue
 
-        LOGGER.info(f"processing stream {stream_id}")
+        LOGGER.info(f"processing stream {table.name}")
 
-        start_time = stream.get_stream_state(stream_id, replication_key) or config_start
+        start_time = (
+            stream.get_stream_state(table.name, replication_key) or config_start
```
Comment (wizhi) [inline tap_salesforce/__init__.py:64]: Just double checking: do we actually use table names for state?

File: tap_salesforce/__init__.py [hunk 3]
```diff
@@ -105,28 +104,26 @@ def main_impl():
...
-            stream_id,
+        for record in sf.get_records(
+            table,
             fields,
-            replication_key,
             start_time,
             end_date=end_time,
             limit=limit,
         ):
-            record = transform_record(raw_record, fields)
```
Comment (wizhi) [inline tap_salesforce/__init__.py:124]: We no longer need to transform the records before writing them?

File: tap_salesforce/__init__.py [hunk 4]
```diff
@@ -105,28 +104,26 @@ def main_impl():
...
-
-            stream.write_record(record, stream_id)
-            state_value = record[replication_key]
 
-            stream.set_stream_state(stream_id, replication_key, state_value)
+            stream.write_record(record, table.name)
+            state_value = datetime.fromisoformat(
+                record[table.replication_key][: -len("+0000")]
+            )
```
Comment (wizhi) [inline tap_salesforce/__init__.py]: Totally dig converting to `datetime` here, but I'd suggest using `datetime.strptime` with the expected format, rather than this. Would make it easier to read.

```suggestion
            state_value = datetime.strptime(record[table.replication_key], "%Y-%m-%dT%H:%M:%S")
```

Alternatively, I see that [`datetime.fromisoformat` should handle the offset by itself.](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromisoformat)

File: tap_salesforce/client.py [hunk 5]
```diff
@@ -13,11 +13,12 @@
     TapSalesforceOauthException,
     TapSalesforceQuotaExceededException,
     TapSalesforceInvalidCredentialsException,
+    QueryLengthExceedLimit,
     build_salesforce_exception,
 )
 from tap_salesforce.metrics import Metrics
 
-MAX_CUSTOM_FIELDS = 450
+MAX_QUERY_LENGTH = 10000
```
Comment (wizhi) [inline tap_salesforce/client.py:21]: Did you verify that they handle 10k length? I thought it was like, 8k.

File: tap_salesforce/client.py [hunk 6]
```diff
@@ -188,25 +179,88 @@ def get_records(
...
+        length = 0
+        index = 0
+        for field in fields:
+            index += 1
+            length += len(field)
+            field_chunk.append(field)
+            if (length > size) or (index == len(fields)):
+                yield field_chunk
+                field_chunk = []
+                length = 0
```
Comment (wizhi) [inline tap_salesforce/client.py:198]: Nice, this is cool.

File: tap_salesforce/client.py [hunk 7]
```diff
@@ -188,25 +179,88 @@ def get_records(
...
+
+    def merge_records(
+        self, paginators: List[Generator[Dict, None, None]], table: Table
+    ) -> Generator[Dict, None, None]:
+        for records in zip(*paginators):
```
Comment (wizhi) [inline tap_salesforce/client.py:203]: Super nice use of `zip`.

File: tap_salesforce/client.py [hunk 8]
```diff
@@ -188,25 +179,88 @@ def get_records(
...
+            merged_record = {}
+            primary_key = None
+            for record in records:
+                if not primary_key:
+                    primary_key = record[table.primary_key]
+                if primary_key != record[table.primary_key]:
+                    raise PrimaryKeyNotMatch(
+                        f"couldn't merge records with different primary keys: {primary_key} and {record[table.primary_key]}"
+                    )
+                merged_record.update(record)
```
Comment (wizhi) [inline tap_salesforce/client.py:213]:
```suggestion
            if not all(records[0][table.primary_key] == record[table.primary_key] for record in records):
                raise PrimaryKeyNotMatch("couldn't merge records with different primary keys")

            yield ChainMap(*records)
```

File: tap_salesforce/client.py [hunk 9]
```diff
@@ -188,25 +179,88 @@ def get_records(
...
+                    "GET",
+                    f"/services/data/{self._API_VERSION}/queryAll/",
+                    params={"q": query},
+                )
+            elif table.primary_key:
+                LOGGER.info(f"query too long {len(query)=}, split into subqueries")
+                paginators = []
+                for field_chunk in self.field_chunker(fields, 8000):
+                    field_chunk.append(table.primary_key)
+                    field_chunk.append(table.replication_key)
```
Comment (wizhi) [inline tap_salesforce/client.py:241]: Wouldn't this cause duplicate fields, since a table likely has its primary- and replication keys in its list of fields already?

File: tap_salesforce/client.py [hunk 10]
```diff
@@ -224,24 +278,26 @@ def get_records(
...
         next_page: Optional[str] = path
         while True:
             resp = self._make_request(method, next_page, data=data, params=params)
 
             resp_data = resp.json()
-
-            yield from resp_data.get("records", [])
-
+            for record in resp_data.get("records", []):
+                yield record
```
Comment (wizhi) [inline tap_salesforce/client.py]: Why the change here?

**Review by wizhi — APPROVED** [review]
Some comments/suggestions.

Otherwise, LGTM.

File: tap_salesforce/__init__.py (hunk 1 above)
Comment (JingLin0) [reply tap_salesforce/__init__.py:57]: The code is trying to get fields for a table to check if the table exist or not. I guess we need to call `/services/data/{self._API_VERSION}/sobjects` endpoint to check if the table exist.

Comment (JingLin0) [reply tap_salesforce/__init__.py:57]: But that need some refactor to make the logic correct. I will leave it as it is for now.

File: tap_salesforce/__init__.py (hunk 2 above)
Comment (JingLin0) [reply tap_salesforce/__init__.py:64]: yes. Here is an example 
```
{
  "bookmarks": {
    "Account": { "LastModifiedDate": "2023-01-25T01:56:41Z" },
    "AccountContactRelation": { "LastModifiedDate": "2023-01-25T00:27:33Z" },
    "AccountHistory": { "CreatedDate": "2023-01-25T01:52:37Z" },
    "Campaign": { "LastModifiedDate": "2023-01-24T23:22:55Z" },
    "CampaignMember": { "LastModifiedDate": "2023-01-25T01:05:40Z" },
    "Contact": { "LastModifiedDate": "2023-01-25T02:01:47Z" },
    "ContactHistory": { "CreatedDate": "2023-01-25T02:01:47Z" },
    "Event": { "LastModifiedDate": "2023-01-24T21:03:52Z" },
    "Lead": { "LastModifiedDate": "2023-01-24T23:45:00Z" },
    "LeadHistory": { "CreatedDate": "2023-01-24T20:30:50Z" },
    "Opportunity": { "LastModifiedDate": "2023-01-25T02:00:11Z" },
    "OpportunityContactRole": { "LastModifiedDate": "2023-01-25T01:08:47Z" },
    "OpportunityFieldHistory": { "CreatedDate": "2023-01-25T01:08:47Z" },
    "OpportunityHistory": { "CreatedDate": "2023-01-25T01:08:49Z" },
    "OpportunityLineItem": { "LastModifiedDate": "2023-01-25T01:08:47Z" },
    "Product2": { "LastModifiedDate": "2022-12-08T13:12:42Z" },
    "RecordType": { "LastModifiedDate": "2023-01-24T21:10:23Z" },
    "Task": { "LastModifiedDate": "2023-01-25T02:04:23Z" },
    "User": { "LastModifiedDate": "2023-01-24T15:51:55Z" },
    "UserRole": { "LastModifiedDate": "2023-01-20T17:57:08Z" }
  }
}
```

File: tap_salesforce/__init__.py (hunk 3 above)
Comment (JingLin0) [reply tap_salesforce/__init__.py:124]: I thought it was stupid that we do any transformation on the raw record. I like to see the data just the way it is.
https://youtu.be/LjhCEhWiKXk

File: tap_salesforce/client.py (hunk 5 above)
Comment (JingLin0) [reply tap_salesforce/client.py:21]: I tested it. It seems fine

File: tap_salesforce/client.py (hunk 9 above)
Comment (JingLin0) [reply tap_salesforce/client.py:241]: If we split the query, the second query won't have primary key and replication_key. I handle deduplication in `construct_query` function

File: tap_salesforce/client.py (hunk 10 above)
Comment (JingLin0) [reply tap_salesforce/client.py]: you are right, I don't have to change here.

File: tap_salesforce/client.py (hunk 8 above)
Comment (JingLin0) [reply tap_salesforce/client.py:213]: I have never used chainmap and all function, it is very cool. Learned something new.

File: tap_salesforce/__init__.py (hunk 4 above)
Comment (JingLin0) [reply tap_salesforce/__init__.py]: it couldn't, because the datetime looks like this: `2023-01-25T02:30:05.000+0000`
